hedera-node/src/main/java/com/hedera/services/store/schedule/HederaScheduleStore.java

```diff
@@ -183,49 +258,239 @@ public boolean isCreationPending() {
+					schedules.get().byEquality().remove(equalityKey);
+				} else {
+					schedules.get().byEquality().put(equalityKey, byEquality);
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean advanceCurrentMinSecond(Instant consensusTime) {
+
+		boolean changed = false;
+
+		long curSecond = schedules.get().getCurrentMinSecond();
+
+		while ((consensusTime.getEpochSecond() > curSecond)
+				&& (!schedules.get().byExpirationSecond().containsKey(new SecondSinceEpocVirtualKey(curSecond)))) {
+
+			++curSecond;
+			changed = true;
+		}
+
+		if (changed) {
+			schedules.get().setCurrentMinSecond(curSecond);
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public List<ScheduleID> nextSchedulesToExpire(Instant consensusTime) {
+
+		long curSecond = schedules.get().getCurrentMinSecond();
+
+		if (!shouldProcessSecond(consensusTime, curSecond)) {
+			return Collections.emptyList();
+		}
+
+		var bySecondKey = new SecondSinceEpocVirtualKey(curSecond);
+
+		var bySecond = schedules.get().byExpirationSecond().get(bySecondKey);
+
+		final List<ScheduleID> list = new ArrayList<>();
+		List<Pair<RichInstant, Long>> toRemove = new ArrayList<>();
+
+		if (bySecond != null) {
+			outer: for (var entry : bySecond.getIds().entrySet()) {
+				var instant = entry.getKey();
+				var ids = entry.getValue();
+				for (int i = 0; i < ids.size(); ++i) {
+					var id = ids.get(i);
+					var scheduleId = EntityNum.fromLong(id).toGrpcScheduleId();
+
+					var schedule = getNoError(scheduleId);
+
+					if (schedule == null) {
+						log.error("bySecond contained a schedule that does not exist! Removing it! second={}, id={}",
+								curSecond, scheduleId);
+						toRemove.add(Pair.of(instant, id));
+
+					} else if (schedule.calculatedExpirationTime().getSeconds() != curSecond) {
+						log.error("bySecond contained a schedule in the wrong spot! Removing and expiring it! spot={}, id={}, schedule={}",
+								curSecond, scheduleId, schedule);
+						toRemove.add(Pair.of(instant, id));
+						list.add(scheduleId);
+
+					} else if (schedule.isDeleted() || schedule.isExecuted()) {
+						list.add(scheduleId);
+
+					} else {
+						break outer;
+					}
+				}
+			}
+
+			if ((!toRemove.isEmpty()) || (bySecond.getIds().size() <= 0)) {
+				bySecond = schedules.get().byExpirationSecond().get(bySecondKey);
+				if (bySecond != null) {
+					bySecond = bySecond.asWritable();
+					for (var p : toRemove) {
+						bySecond.removeId(p.getKey(), p.getValue());
+					}
+
+					if (bySecond.getIds().size() <= 0) {
+						log.error("bySecond was unexpectedly empty! Removing it! second={}", curSecond);
+						schedules.get().byExpirationSecond().remove(bySecondKey);
+					} else {
+						schedules.get().byExpirationSecond().put(bySecondKey, bySecond);
+					}
+				}
+			}
+		}
+
+		return list;
+	}
+
+	@Override
+	@Nullable
+	public ScheduleID nextScheduleToEvaluate(Instant consensusTime) {
+
+		long curSecond = schedules.get().getCurrentMinSecond();
```

Comment:
This setup duplicates that of `nextSchedulesToExpire`; an internal `@Nullable ScheduleSecondVirtualValue nextSecondToProcess(final Instant now)` would save a few lines.

Reply:
I'll refactor a little time permitting ;-) Trying to get everything fixed first.

Reply:
I usually avoid helper methods when they must return multiple values and the usages are slightly different. In this example it would need to return 3 different values and if we ever get getForModify working then the usages might be very different.